### PR TITLE
[quant][graphmode][refactor] Identify weights and bias by argument position in aten call

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29147 [quant][graphmode][refactor] Identify weights and bias by argument position in aten call**

Summary:
Previously we use a vector of weight and bias to record the values of weight/bias and we
assume we'll get them by GetAttr nodes, then we propagate these values through the function calls

However, it doesn't work if we also do some transformations on these values right now, we'll need
to mark all the values that's produced by weight/bias as weight/bias, e.g.
```
%w = GetAttr[name="weight"](%conv)
%wt = aten::transpose(%w)
%r = aten::conv2d(..., %wt, ...)
```
we'll mark both %w and %wt as weight. This is a bit over compilicated to support this.

Alternatively, we can identify weights by argument positions, e.g.
for call %r = aten::conv2d(..., %w, ...), we know the argument 1 is weight, argument 2 is bias.

Test Plan:
test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18362839](https://our.internmc.facebook.com/intern/diff/D18362839)

### CODE REVIEW

**Review by ZolotukhinM — CHANGES_REQUESTED** [review]
In general I think it's a good change, but it needs some cleanup before it can go in.

For instance, the current implementation of `insertObserverFor` creates an impression that it only handles "conv2d" and "linear". This impression is wrong, as we *can* handle all other quantizable functions here as well, but at the moment 50% of the code is dealing with those two functions. It needs to be refactored in such a way that the logic is clear (compare for instance with the original code of this function).

File: torch/csrc/jit/passes/quantization.cpp [hunk 1]
```diff
@@ -160,17 +149,31 @@ Node* InsertObserversHelper::insertObserverFor(
     Graph* g,
     script::Module& module,
     const QConfig& qconfig) {
-  // Skip observing bias
-  if (bias_values_.count(v)) {
-    return nullptr;
+  for (const Use& u : v->uses()) {
+    if (u.user->kind() == Symbol::aten("conv2d")) {
+      if (v == u.user->inputs().at(2)) {
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp]: will add a new functions for these checks

File: torch/csrc/jit/passes/quantization.cpp [hunk 2]
```diff
@@ -130,47 +130,67 @@ class InsertObserversHelper {
...
+      if (func_name == "linear" && v == u.user->inputs().at(3)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+bool isWeightOfConvOrLinear(Value* v) {
+  for (const Use& u : v->uses()) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Do we actually allow multiple uses? How would we quantize it if it's used in several places? If we don't support this case, we should assert that there is only one use.

**Review by ZolotukhinM — APPROVED** [review]
Thanks for following up on the remarks! It looks good!

File: torch/csrc/jit/passes/quantization.cpp (hunk 2 above)
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp]: we don't support multiple uses, and for weight we probably don't need to. I'll add an assert.

**Pull request closed without merging**